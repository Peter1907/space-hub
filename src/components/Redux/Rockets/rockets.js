import { createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const GET = './Rockets/rockets/GET';
const UPDATE = './Rockets/rockets/UPDATE';
const GET_STORED = './Rockets/rockets/GET_STORED';

const rocketsReducer = (state = [], action) => {
  switch (action.type) {
    case `${GET}/fulfilled`:
      const DATA = action.payload.map((item) => ({
        id: item.id,
        name: item.rocket_name,
        description: item.description,
        image: item.flickr_images[0],
        reserved: false,
      }));
      localStorage.setItem('ROCKET_DATA', JSON.stringify(DATA));
      return DATA;
    case GET_STORED:
      return action.payload;
    case UPDATE:
      const modifiedState = state.map((item) => {
        if (item.id === action.id) {
          return ({
            ...item,
            reserved: !item.reserved,
          });
        }
        return item;
      })
      localStorage.setItem('ROCKET_DATA', JSON.stringify(modifiedState));
      return modifiedState;
    default:
      return state;
  }
};

const getRockets = createAsyncThunk(
  GET,
  async () => {
    const response = await axios('https://api.spacexdata.com/v3/rockets');
    return response.data; 
  }
);

const getStored = () => ({
  type: GET_STORED,
  payload: JSON.parse(localStorage.getItem('ROCKET_DATA')),
});

const updateStatus = (id) => ({
  type: UPDATE,
  id,
});

export default rocketsReducer;
export { getRockets, updateStatus, getStored };
