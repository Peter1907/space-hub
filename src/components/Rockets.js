/* eslint-disable react-hooks/exhaustive-deps */
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { getRockets, getStored, updateStatus } from './Redux/Rockets/rockets';
import styles from './Rockets.module.css';

export default function Rockets() {
  const dispatch = useDispatch();
  const rockets = useSelector((state) => state.rockets);
  const DATA = localStorage.getItem('ROCKET_DATA');

  const getData = () => (DATA ? dispatch(getStored()) : dispatch(getRockets()));

  useEffect(() => {
    getData();
  }, []);

  return (
    <div className={styles.rockets}>
      {rockets.map((item) => (
        <Rocket
          key={item.id}
          id={item.id}
          name={item.name}
          description={item.description}
          img={item.image}
          reserved={item.reserved}
        />
      ))}
    </div>
  );
}

function Rocket({
  id,
  name,
  description,
  img,
  reserved,
}) {
  const dispatch = useDispatch();

  return (
    <div className={styles.rocket}>
      <img src={img} alt={`The ${name} Rocket`} />
      <div className={styles.info}>
        <h2>{name}</h2>
        <p>
          {reserved && <span className={styles.tag}>Reserved</span>}
          {description}
        </p>
        {!reserved
          && (
            <button
              type="button"
              className={styles.reservationBtn}
              onClick={() => dispatch(updateStatus(id))}
            >
              Reserve Rocket
            </button>
          )}
        {reserved
          && (
            <button
              type="button"
              className={styles.cancelationBtn}
              onClick={() => dispatch(updateStatus(id))}
            >
              Cancel Reservation
            </button>
          )}
      </div>
    </div>
  );
}

Rocket.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  img: PropTypes.string.isRequired,
  reserved: PropTypes.bool.isRequired,
};
