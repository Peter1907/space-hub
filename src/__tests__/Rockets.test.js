/* eslint-disable*/

import renderer from 'react-test-renderer';
import Rockets from '../__mocks__/RocketsMock';

it('renders correctly', () => {
  const tree = renderer
    .create(<Rockets />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
