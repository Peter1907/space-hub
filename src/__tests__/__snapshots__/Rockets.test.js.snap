// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="rockets"
>
  <div
    className="rocket"
  >
    <img
      alt="The Falcon 1 Rocket"
      src="https://imgur.com/DaCfMsj.jpg"
    />
    <div
      className="info"
    >
      <h2>
        Falcon 1
      </h2>
      <p>
        <span
          className="tag"
          style={
            Object {
              "display": "none",
            }
          }
        >
          Reserved
        </span>
        The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.
      </p>
      <button
        style={
          Object {
            "backgroundColor": "#007bff",
            "border": "none",
            "color": "white",
          }
        }
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
  <div
    className="rocket"
  >
    <img
      alt="The Falcon 9 Rocket"
      src="https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg"
    />
    <div
      className="info"
    >
      <h2>
        Falcon 9
      </h2>
      <p>
        <span
          className="tag"
          style={
            Object {
              "display": "none",
            }
          }
        >
          Reserved
        </span>
        Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.
      </p>
      <button
        style={
          Object {
            "backgroundColor": "#007bff",
            "border": "none",
            "color": "white",
          }
        }
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
  <div
    className="rocket"
  >
    <img
      alt="The Falcon Heavy Rocket"
      src="https://farm5.staticflickr.com/4599/38583829295_581f34dd84_b.jpg"
    />
    <div
      className="info"
    >
      <h2>
        Falcon Heavy
      </h2>
      <p>
        <span
          className="tag"
          style={
            Object {
              "display": "none",
            }
          }
        >
          Reserved
        </span>
        With the ability to lift into orbit over 54 metric tons (119,000 lb)--a mass equivalent to a 737 jetliner loaded with passengers, crew, luggage and fuel--Falcon Heavy can lift more than twice the payload of the next closest operational vehicle, the Delta IV Heavy, at one-third the cost.
      </p>
      <button
        style={
          Object {
            "backgroundColor": "#007bff",
            "border": "none",
            "color": "white",
          }
        }
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
  <div
    className="rocket"
  >
    <img
      alt="The Starship Rocket"
      src="https://live.staticflickr.com/65535/48954138962_ee541e6755_b.jpg"
    />
    <div
      className="info"
    >
      <h2>
        Starship
      </h2>
      <p>
        <span
          className="tag"
          style={
            Object {
              "display": "none",
            }
          }
        >
          Reserved
        </span>
        Starship and Super Heavy Rocket represent a fully reusable transportation system designed to service all Earth orbit needs as well as the Moon and Mars. This two-stage vehicle — composed of the Super Heavy rocket (booster) and Starship (ship) — will eventually replace Falcon 9, Falcon Heavy and Dragon.
      </p>
      <button
        style={
          Object {
            "backgroundColor": "#007bff",
            "border": "none",
            "color": "white",
          }
        }
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
</div>
`;
